commit 76c50b9eea2dd97d83dbf5fa089546e2ee70ac68 (refs/changes/23/5623/1)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2019-02-22T02:06:37+00:00 (8 months ago)
    
    backout: OS-7566 Clean up qemu use of old CTF tools (breaks CTF object data)

diff --git a/Makefile.target b/Makefile.target
index 2179d67..2050146 100644
--- a/Makefile.target
+++ b/Makefile.target
@@ -382,8 +382,9 @@ ifeq ($(TRACE_BACKEND),dtrace)
 ifneq ($(strip $(CONFIG_SOLARIS)),)
 $(QEMU_PROG): $(obj-y) $(obj-$(TARGET_BASE_ARCH)-y)
 	$(call quiet-command, dtrace $(CONFIG_DTRACE_FLAGS) -o ../trace-dtrace.o -s ../trace-dtrace.dtrace -G $(obj-y) $(obj-$(TARGET_BASE_ARCH)-y),"  LINK  $(TARGET_DIR)$@.dtrace")
+	$(call CTFCONVERT_CMD)
 	$(call LINK,$(obj-y) $(obj-$(TARGET_BASE_ARCH)-y) ../trace-dtrace.o)
-	$(call CTFCONVERT_CMD, $(obj-y) $(obj-$(TARGET_BASE_ARCH)-y))
+	$(call CTFMERGE_CMD, $(obj-y) $(obj-$(TARGET_BASE_ARCH)-y))
 else
 $(QEMU_PROG): $(obj-y) $(obj-$(TARGET_BASE_ARCH)-y)
 	$(call LINK,$(obj-y) $(obj-$(TARGET_BASE_ARCH)-y))
diff --git a/ctf.sh b/ctf.sh
new file mode 100755
index 0000000..11fc139
--- /dev/null
+++ b/ctf.sh
@@ -0,0 +1,29 @@
+#!/usr/bin/bash
+#
+# We need to run ctfconvert on all the .o files in qemu. However, some of these
+# .o files contain some snippets that are going to cause ctfconvert to fail. If
+# ctfconvert is run with the -i option, it will delete the .o file. This is bad.
+# Instead we end up using a temporary file and move over it. 
+#
+
+sh_arg0=$(basename $0)
+
+function fail
+{
+        local msg="$*"
+        [[ -z "$msg" ]] && msg="failed"
+        echo "$sh_arg0: $msg" >&2
+        exit 1
+}
+
+[[ $# -eq 1 ]] || fail "missing arguments"
+
+# CTFCONVERT may contain a wildcard, so expand it out:
+ctfconvert=$(echo ${CTFCONVERT})
+[[ -x ${ctfconvert} ]] || fail "could not find ctfconvert at $CTFCONVERT"
+
+echo "Converting $1"
+$ctfconvert -L VERSION -o $1.ctf $1
+[[ $? -ne 0 ]] && exit 1
+mv $1.ctf $1
+exit 0
diff --git a/rules.mak b/rules.mak
index 73ca4e8..81bfc03 100644
--- a/rules.mak
+++ b/rules.mak
@@ -25,7 +25,9 @@ QEMU_DGFLAGS += -MMD -MP -MT $@ -MF $(*D)/$(*F).d
 
 LINK = $(call quiet-command,$(CC) $(QEMU_CFLAGS) $(CFLAGS) $(LDFLAGS) -o $@ $(1) $(LIBS),"  LINK  $(TARGET_DIR)$@")
 
-CTFCONVERT_CMD = $(call quiet-command, $(CTFCONVERT) -L VERSION -o $@ $@,"  CTFCONVERT  $(TARGET_DIR)$@")
+CTFMERGE_CMD = $(call quiet-command, $(CTFMERGE) -L VERSION -o $@ $(1),"  CTFMERGE  $(TARGET_DIR)$@")
+
+CTFCONVERT_CMD = $(call quiet-command, CTFCONVERT=$(CTFCONVERT) find ../ -type f -name '*.o' -exec ../ctf.sh '{}' \;)
 
 ifeq ($(TRACE_BACKEND),dtrace)
 ifneq ($(strip $(CONFIG_SOLARIS)),)
